### PR TITLE
dmtcp_restart: Add a debug option to enable auto gdb attach

### PR DESCRIPTION
**Pull request by rohgarg**

### CODE REVIEW

**Comment by rohgarg** [comment]
@karya0: Could you please review this? This is still not that polished; posting this for feedback/comments.

File: src/pluginmanager.cpp [hunk 1]
```diff
@@ -152,6 +152,14 @@ void PluginManager::processRestartBarriers()
 {
   PluginManager::registerBarriersWithCoordinator();
 
+  /* FIXME(rohgarg):
+   *  The following write can return -1 if PROTECTED_DEBUG_PIPE_FD is uninitialized.
+   *  A proper solution would require the ability to share state between dmtcp_restart
+   *  and this file.
+   */
+  int currentDebugLevel = 6;
+  write(PROTECTED_DEBUG_PIPE_FD, &currentDebugLevel, sizeof(currentDebugLevel));
```
Comment (karya0) [inline src/pluginmanager.cpp]: Use  `Util::isValidFd(PROTECTED_DEBUG_PIPE_FD)` to check for fd validity.

File: src/threadlist.cpp [hunk 2]
```diff
@@ -673,6 +673,14 @@ void ThreadList::postRestart(void)
 
   restoreInProgress = true;
 
+  /* FIXME(rohgarg):
+   *  The following write can return -1 if PROTECTED_DEBUG_PIPE_FD is uninitialized.
+   *  A proper solution would require the ability to share state between dmtcp_restart
+   *  and this file.
+   */
+  int currentDebugLevel = 7;
+  write(PROTECTED_DEBUG_PIPE_FD, &currentDebugLevel, sizeof(currentDebugLevel));
```
Comment (karya0) [inline src/threadlist.cpp]: ditto!

**Comment by rohgarg** [comment]
I have now updated the PR based on feedback from @gc00 and @karya0.

File: include/protectedfds.h [hunk 3]
```diff
@@ -43,6 +43,7 @@ enum ProtectedFds {
   PROTECTED_READLOG_FD,
   PROTECTED_ENVIRON_FD,
   PROTECTED_NS_FD,
+  PROTECTED_DEBUG_PIPE_FD,
```
Comment (karya0) [inline include/protectedfds.h]: s/PIPE/SOCKET

File: src/dmtcp_restart.cpp [hunk 4]
```diff
@@ -96,6 +96,8 @@ static const char* theUsage =
   "              (default: use the same directory used in previous checkpoint)\n"
   "  --tmpdir PATH (environment variable DMTCP_TMPDIR)\n"
   "              Directory to store temporary files (default: $TMDPIR or /tmp)\n"
+  "  --gdb    LEVEL\n"
```
Comment (karya0) [inline src/dmtcp_restart.cpp]: Let's not document this for now until we have had enough test coverage

File: src/dmtcp_restart.cpp [hunk 5]
```diff
@@ -393,6 +398,43 @@ static void runMtcpRestart(int is32bitElf, int fd, ProcessInfo *pInfo)
...
+      close(debugPipe[1]);
+      do {
+        rc = read(debugPipe[0], &currentDebugLevel, sizeof(currentDebugLevel));
+        if (rc < 0) break;
+        rc = write(debugPipe[0], &requestedDebugLevel, sizeof(currentDebugLevel));
+        if (rc < 0) break;
+      } while (currentDebugLevel != requestedDebugLevel);
+      if (rc < 0) {
+        int status;
+        waitpid(pid, &status, 0);
```
Comment (karya0) [inline src/dmtcp_restart.cpp]: Shouldn't we exit here?

File: src/dmtcp_restart.cpp [hunk 6]
```diff
@@ -393,6 +398,43 @@ static void runMtcpRestart(int is32bitElf, int fd, ProcessInfo *pInfo)
...
+                               const_cast<char*>(pInfo->procSelfExe().c_str()),
+                               cpid,
+                               NULL};
+      execvp(command[0], command);
+    } else if (pid == 0) {
+      close(debugPipe[0]); // child doesn't need the read end
+      JASSERT(dup2(debugPipe[1], PROTECTED_DEBUG_PIPE_FD) == PROTECTED_DEBUG_PIPE_FD)(JASSERT_ERRNO);
+      close(debugPipe[1]);
+#ifdef HAS_PR_SET_PTRACER
+      prctl(PR_SET_PTRACER, PR_SET_PTRACER_ANY, 0, 0, 0); // Allow 'gdb attach'
```
Comment (karya0) [inline src/dmtcp_restart.cpp]: I don't think we need this because the restarted process is a direct child.

**Pull request merged**